// @nervina-labs/cota-sdk/lib/service/cota/isssuer.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateIssuerInfoTx = void 0;
const constants_1 = require("../../constants");
const utils_1 = require("../../utils");
const generateIssuerMetadata = (issuerInfo) => {
    const issuerMeta = {
        id: 'CTMeta',
        ver: '1.0',
        metadata: {
            target: 'output#0',
            type: 'issuer',
            data: Object.assign({ version: '0' }, issuerInfo),
        },
    };
    return (0, utils_1.append0x)((0, utils_1.utf8ToHex)(JSON.stringify(issuerMeta)));
};
const generateIssuerInfoTx = (service, cotaLock, issuerInfo, fee = constants_1.FEE, isMainnet = false) => __awaiter(void 0, void 0, void 0, function* () {
    const cotaType = (0, constants_1.getCotaTypeScript)(isMainnet);
    const cotaCells = yield service.collector.getCells(cotaLock, cotaType);
    if (!cotaCells || cotaCells.length === 0) {
        throw new Error("Cota cell doesn't exist");
    }
    const cotaCell = cotaCells[0];
    const inputs = [
        {
            previousOutput: cotaCell.outPoint,
            since: '0x0',
        },
    ];
    const outputs = [cotaCell.output];
    outputs[0].capacity = `0x${(BigInt(outputs[0].capacity) - fee).toString(16)}`;
    const cotaOutput = yield service.collector.getLiveCell(cotaCell.outPoint);
    const outputsData = [cotaOutput.data.content];
    const cellDeps = [(0, constants_1.getCotaCellDep)(isMainnet)];
    const rawTx = {
        version: '0x0',
        cellDeps,
        headerDeps: [],
        inputs,
        outputs,
        outputsData,
        witnesses: [],
    };
    rawTx.witnesses = rawTx.inputs.map((_, i) => i > 0 ? '0x' : { lock: '', inputType: '', outputType: generateIssuerMetadata(issuerInfo) });
    return rawTx;
});
exports.generateIssuerInfoTx = generateIssuerInfoTx;
